### PR TITLE
PR: Improve CFI 2017 and TM30-18 Performance

### PR DESCRIPTION
**Pull request by tjdcs**
# Summary

This PR implements several performance improvements, again, that should improve the speed and quality of the entire code base. These improvements mainly revolve around caching in certain places during sd_to_XYZ in order to prevent excessive copies of CMFs etc... 

Because these features change caching behavior there is a small risk that there will be regressions, despite the tests passing so it is requested that it is tried in some other developer applications before merging. 

2X performance improvement in CFI 2017 workloads

# Preflight

**Code Style and Quality**

- [x] Unit tests have been implemented and passed.
- [x] Pyright static checking has been run and passed.
- [x] Pre-commit hooks have been run and passed.
- [ ] ~New transformations have been added to the *Automatic Colour Conversion Graph*.~
- [ ] ~New transformations have been exported to the relevant namespaces, e.g. `colour`, `colour.models`.~

**Documentation**

- [x] ~New features are documented along with examples if relevant.~
- [x] ~The documentation is [Sphinx](https://www.sphinx-doc.org/en/master/) and [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant.~

### CODE REVIEW

**Comment by tjdcs** [comment]
Looks like I did actually mess up the tests some how... I was really hoping that was my local environment quarks.

**Comment by tjdcs** [comment]
@KelSolaar looks like the only issue left is the PCA issue. We'll have to figure out a plan for those PCA tests. FYI three of them are failing on my local machine. I suspect that small changes in the code base are affecting floating point errors and that's feeding forward into PCA results.

**Comment by tjdcs** [comment]
Status: around 42x faster than v.0.4.2 release

**Comment by tjdcs** [comment]
@KelSolaar we have about an 80x performance improvement for TM30 in this branch now. I think it should be merged. I'd like to get another 3x faster than this... but that will require more significant refactoring. First targets if anyone wants to pick it up are the Ohno 2013 code and the CAM code... 

I might introduce / write a paper on a highly compute optimized CAM. That could be nice in the future. 

Anyway... The only failing tests look like some of those precision errors. The tests pass on my local machine. Let me know if you want me to fix the tests, or what you'd like to do to pass this. 

And of course... looking forward to your review.

File: colour/temperature/ohno2013.py
Comment (KelSolaar) [inline colour/temperature/ohno2013.py:184]: **Minor**

An example would not hurt.

File: colour/temperature/tests/test_ohno2013.py
Comment (KelSolaar) [inline colour/temperature/tests/test_ohno2013.py:116]: **Minor**

We should probably use 0.05 here which is the limit at which the concept of CCT is not valid anymore.

File: colour/temperature/tests/test_ohno2013.py
Comment (KelSolaar) [inline colour/temperature/tests/test_ohno2013.py:118]: **Minor**

`D_uv` for consistency?

File: colour/temperature/ohno2013.py
Comment (KelSolaar) [inline colour/temperature/ohno2013.py:298]: **Pedantry**

Ahaha :D

File: colour/utilities/array.py
Comment (KelSolaar) [inline colour/utilities/array.py:1919]: **Discussion/Minor**

I wish Numpy offered a native way to hash ndarrays, given we do that often, we should maybe write a small wrapper function to do that.

`shape` is a tuple btw, so you don't need to convert it:

```
>>> hash((1, 2, 3))
529344067295497451
```

File: colour/utilities/array.py
Comment (KelSolaar) [inline colour/utilities/array.py:2073]: **Pedantry**

Unneeded break line removal.

File: colour/algebra/common.py
Comment (KelSolaar) [inline colour/algebra/common.py:88]: **Minor**

This is probably incorrectly formatted!?

File: colour/temperature/tests/test_ohno2013.py
Comment (KelSolaar) [inline colour/temperature/tests/test_ohno2013.py:80]: **Minor**

We prefer having a class per-definition.

File: colour/temperature/ohno2013.py
Comment (KelSolaar) [inline colour/temperature/ohno2013.py:93]: **Nice/Minor**

Quite the radical change!

The reported default does not seem to be correct.

File: colour/temperature/ohno2013.py
Comment (KelSolaar) [inline colour/temperature/ohno2013.py:206]: **Minor**

Ditto for example.

File: .gitignore
Comment (KelSolaar) [inline .gitignore:24]: **Minor**

Noeeeeeeeezzz :(

**Review by KelSolaar — CHANGES_REQUESTED** [review]
Thanks a lot @tjdcs!

A lot of good and exciting stuff there. I did a first review pass. I would be keen to clean up the commits a bit as discussed on Slack.

File: .gitignore
Comment (tjdcs) [reply .gitignore:24]: No idea why this happened. :/ but I will revert before merge. So gross. I hate it too 😃

File: colour/appearance/ciecam02.py
Comment (tjdcs) [reply colour/appearance/ciecam02.py:625]: Yeah I was going to go through and put some of these back in where I moved them. there is one relatively important one I want to remove because the values that can cause an issue are generated in the same function context and can be made safer.

File: colour/characterisation/datasets/colour_checkers/chromaticity_coordinates.py
Comment (tjdcs) [reply colour/characterisation/datasets/colour_checkers/chromaticity_coordinates.py:484]: I touched this file and my IDE changed it. There’s going to be a lot of these totally unintentional

File: colour/colorimetry/blackbody.py
Comment (tjdcs) [reply colour/colorimetry/blackbody.py:133]: This should elide a copy I think… I did test and get a performance improvement. But I should double check when I fix the other comments.

File: colour/colorimetry/spectrum.py
Comment (tjdcs) [reply colour/colorimetry/spectrum.py:715]: Will revert. This was entirely to get around not having the call back feature.

File: colour/continuous/multi_signals.py
Comment (tjdcs) [reply colour/continuous/multi_signals.py:398]: Yes. Avoiding an extra copy. Stacking the ranges is really slow before hashing. And accessing values does the same operation. Calling range on every sub signal might be an option that I missed the first time.

File: colour/models/rgb/rgb_colourspace.py
Comment (tjdcs) [reply colour/models/rgb/rgb_colourspace.py:696]: Yes we should initialize to none explicitly. Good review point. This was entirely because I kept getting frustrated with load times 😂 there are a lot of RGB spaces that get initialized the first time colour is imported.

File: colour/plotting/tm3018/tests/test_report.py
Comment (tjdcs) [reply colour/plotting/tm3018/tests/test_report.py:136]: Yeah. Plus this is broken right now. I will remove before final review. But it’s an idea that we should have some discussion about.  I’ll make a note about why I noticed I needed this when coding and open an issue.

File: colour/quality/cfi2017.py
Comment (tjdcs) [reply colour/quality/cfi2017.py:95]: Because I vecorized the cam call I have to I stack the values, then use the iterate operator to turn arrays into args for this data object constructor. It was around 5% or run time when I removed it. :/

File: colour/quality/cfi2017.py
Comment (tjdcs) [reply colour/quality/cfi2017.py:463]: Too slow 😭

my highly optimized matlab implementation is about 3x faster than this but makes many architectural compromises that colour provides. I don’t expect to ever reach the same performance, but this was one of those changes that helped me get above 500 evaluations per second on my MBP. 

Definitely a consistency compromise… :( but could be clarified with a comment.

File: colour/quality/tests/test_cqs.py
Comment (tjdcs) [reply colour/quality/tests/test_cqs.py:38]: Because of the changes to the plankian code we had some very minor value changes in all quality codes.

File: colour/quality/tm3018.py
Comment (tjdcs) [reply colour/quality/tm3018.py:177]: Clever implicit expansion / broadcasting is the most powerful performance tool we have in numpy code. It can really cleverly remove the need for tstack and unstack in other places.  But it’s obviously much harder to read especially for people who haven’t read the sacred texts.

An aerospace simulations engineer taught me these. Fun stuff.

File: colour/temperature/tests/test_ohno2013.py
Comment (tjdcs) [reply colour/temperature/tests/test_ohno2013.py:116]: Sadly, Ohno’s method is actually only accurate to +/- 1K for values less than 0.003 when the table has a spacing of 1.01. I had to improve the spacing and write my own correction function to extend the accuracy even to this level. 

It is, unfortunately(?) only an approximation. And of course, a more dense table is also slower.

File: colour/colorimetry/spectrum.py
Comment (tjdcs) [reply colour/colorimetry/spectrum.py]: Screen Shot 2023-05-03 at 11 25 32 PM

These tests were failing. I'm not sure why. It's a strange case where the class gets initialized with zero domain at some point... Could be that there is another issue. but I just wanted to make the red checks turn green.

**Comment by tjdcs** [comment]
@KelSolaar This is greatly cleaned up now if you want to give it another pass.

File: colour/appearance/ciecam02.py
Comment (tjdcs) [reply colour/appearance/ciecam02.py]: Reverted

File: colour/colorimetry/spectrum.py
Comment (tjdcs) [reply colour/colorimetry/spectrum.py:2838]: I had to pump up my line removals in the PR. It's good for the profile numbers.

File: colour/plotting/tm3018/tests/test_report.py
Comment (tjdcs) [reply colour/plotting/tm3018/tests/test_report.py]: Reverted. SHould revisit in the future.

File: colour/algebra/common.py
Comment (tjdcs) [reply colour/algebra/common.py]: I don't see this change anymore, but I did clean this up with a type alias

**Comment by tjdcs** [comment]
ad67a6b2c8ff2f0907126ed63dc33020e200b06c adds a dependency on xxhash which is an extremely fast hashing algorithm specifically suited for dictionary keys. https://github.com/Cyan4973/xxHash#quality

**Comment by KelSolaar** [comment]
@tjdcs : Where are we with this one?

**Comment by tjdcs** [comment]
Bizarre corner cases in the initialization of `SpectralDistribution` and `Signal` appear to be failing tests. And there is a mountain of type errors that I can't wade through because my development environment and the CI environment are different. CI complains there are only a few type errors, but pyright on my machine reports more than 600. 

It's going to be a small while before I can address these on my own. :/ i have to prep for cinegear, and then infocomm. I have urgent work projects taking the few days I have in between. 

Probably the end of June is when I will have serious time to sit down with this again.

**Comment by KelSolaar** [comment]
Thanks, sounds good! I'm going to update Pyright on our side so that we run the latest and then we can rebase your branch on top of `develop`.

**Comment by KelSolaar** [comment]
`develop` is up-to-date against the latest `pyright` and a few fixes have been rolled for static analysis.

**Comment by KelSolaar** [comment]
I looked quickly why the docs build is failing and I think that we are destroying the blackbody spectral distributions that we defined here : https://github.com/colour-science/colour/blob/36d925516f8716750a609da32bc45b0c116c2244/utilities/generate_plots.py#L187, there: https://github.com/colour-science/colour/blob/36d925516f8716750a609da32bc45b0c116c2244/colour/colorimetry/tristimulus_values.py#L858

This is one of the reason why `reshape_sd` was always operating on a copy, now we have introduced those kind of hard to debug side effects. I think we need to use a context manager or expose the copy argument in the definitions that use `reshape_sd`. Food for thoughts!

I would be keen to drop the Poetry commit from this PR as it is a separate problem and just created a huge mess on my end, took me a while to realise what had happened.

**Comment by tjdcs** [comment]
I think I rebased on that branch. I can drop and rebase to developer again, but I needed it because I wanted to depend on both this branch and that update for a workbench project. 

Understood about it being a mess though...

There are several other tests failing in the debug module examples tests. I did finally get the rest of my toolchain working like expected so I should be able to get farther along with this. It looks like the examples that are failing are doing so because of the ndcopy array stuff... I'll dig in more tonight.

**Comment by KelSolaar** [comment]
I was incorrect WRT to the copy and found out what caused the documentation to break. As we changed the spectral shape here from `(0, 10000, 1)` to `(1, 10001, 1)` it caused the integration to fail because the shape become unaligned:

image

The logic in`colour.colorimetry.sd_to_XYZ_tristimulus_weighting_factors_ASTME308` definition does not work properly:

```python
    if sd.shape.boundaries != cmfs.shape.boundaries:
        runtime_warning(
            f'Trimming "{illuminant.name}" spectral distribution shape to '
            f'"{cmfs.name}" colour matching functions shape.'
        )
        sd = reshape_sd(sd, cmfs.shape, "Trim", copy=False)
```

We check if the boundaries are different but do not account for interval difference nor the fact that it could be misaligned, which is what is happening here:

`(360, 780, 1) <--> (1, 10001, 1)`

I think we might need to `interpolate` / `align` rather than `trim`.

**Comment by KelSolaar** [comment]
Oh, Just saw your commit, I'm trying to come up with something that will not break unit tests, I realised that aligning unfortunately does break and kind of defeats the idea of using the weighting factors in the first place. We cannot trim predictably a spectral distribution whose start or end wavelength modulus to 10 is not 0. Simple repro case:

```python
from colour import sd_ones, SpectralShape, sd_to_XYZ

o = 1
sd_to_XYZ(sd_ones(SpectralShape(o, 1000 + o, 20)))
```

Coming up with something soon!

**Comment by tjdcs** [comment]
@KelSolaar ready for a final review, assuming that you are OK if I hurt coverage by 3 lines...

**Comment by KelSolaar** [comment]
Just pushed a small varnishing pass, hopefully we can merge right after if everything builds.

**Comment by KelSolaar** [comment]
LGTM, happy to merge?

**Comment by tjdcs** [comment]
Make it so!

**Comment by KelSolaar** [comment]
Merged, thanks for the great work @tjdcs! Do you remember roughly the performance improvements for the release notes?

**Comment by tjdcs** [comment]
45x from pre-merge commit. From when the work started I think it was around 100x I can check that tomorrow. Some other performance stuff got merged in early or there were other changes. 

I backed out some of the hashing improvements which is the next big thing, but in order to do it with good code style we need to make xxhash a default install. (need to access private `signal` variables in `multisignal` `__hash__`)

**Comment by tjdcs** [comment]
`colour_fidelity_index_ANSIIESTM3018` 100x compared to v0.4.2

`uv_to_CCT_Ohno2013` 5x faster thank v.0.4.3 which might also be of interest in a release note

**Pull request merged**
